Implementation based on the dbus-python library

linrunner opened the pull request:
Status: 
* Covers all Methods/Signals/Properties ppd offers
* Working and tested with GNOME 49 desktop

Rationale for the switch to dbus-python:
* pydbus is unmaintained since 2018 [2], posing an incalculable
  risk to the sustainability of TLP-PD.
* dbus-python is maintained to the present [1]. Although it is called
  "legacy" and "deprecated" in [3], it is used by tuned-ppd [4],
  for example.
* Of those listed in [3], only dasbus and python-sdbus are still
  maintained. Unfortunately, the latter is only packaged in a few
  distributions.
* Conclusion: as of today, the extent of dbus-python appears to be
  sufficient for TLP-PD, so the conservative approach was chosen.

Other changes:
* Rename tlp-pd to tlp-pd.in to enable Makefile parameter substitution

Open items:
* Command line utility to control the daemon via D-Bus
  (similar to powerprofilesctl)

References:
    * [1] https://gitlab.freedesktop.org/dbus/dbus-python/
    * [2] https://github.com/LEW21/pydbus/
    * [3] https://wiki.python.org/moin/DbusExamples
    * [4] https://github.com/redhat-performance/tuned/blob/master/tuned-ppd.py

linrunner committed:
Reimplement based on the dbus-python library

Status: MVP working and tested with GNOME 49 desktop

Rationale for the switch to dbus-python:
* pydbus is unmaintained since 2018 [2], posing an incalculable
  risk to the sustainability of TLP-PD.
* dbus-python is maintained to the present [1]. Although it is called
  "legacy" and "deprecated" in [3], it is used by tuned-ppd [4],
  for example.
* Of those listed in [3], only dasbus and python-sdbus are still
  maintained. Unfortunately, the latter is only packaged in a few
  distributions.
* Conclusion: as of today, the extent of dbus-python appears to be
  sufficient for TLP-PD, so the conservative approach was chosen.

Other changes:
* Rename tlp-pd to tlp-pd.in to enable Makefile parameter substitution

Open items:
* The daemon does not immediately detect changes to the profile made
  in the background via tlp <profile>. This means that the desktop
  does not notice the change either.
* Debug logging and command line switch
* Polkit concept
* D-BUS introspection does not expose all methods/properties
* README, manpages
* Command line utility to control the daemon via D-Bus
  (similar to powerprofilesctl)
* Assess the actual need for:
  * Methods HoldProfile/ReleaseProfile
  * Legacy bus name net.hadess.PowerProfiles

References:
* [1] https://gitlab.freedesktop.org/dbus/dbus-python/
* [2] https://github.com/LEW21/pydbus/
* [3] https://wiki.python.org/moin/DbusExamples
* [4] https://github.com/redhat-performance/tuned/blob/master/tuned-ppd.py

linrunner committed:
Use syslog for logging

linrunner committed:
TLP notifies all changes to its active profile

Add method SyncProfile to receive profile change notifications from TLP.
Change self._active_profile when TLP notifies instead of directly after
Set(ActiveProfile).

linrunner committed:
comment

linrunner committed:
HoldProfile: let TLP notification change self._active_profile

linrunner committed:
Fix GetAll.ActiveProfile

linrunner committed:
Add command option --debug|-D

linrunner committed:
Run tlp <profile> in detached mode

Rationale: avoid deadlocks in the event loop caused by tlp
invoking SyncProfile.

linrunner committed:
Log profile at start

tbodt commented on tlp-pd.in:
```diff
@@ -0,0 +1,394 @@
+                name="org.freedesktop.DBus.Error.UnknownInterface",
+            )
+
+        return {
```
I'd suggest making this a series of self._get_property calls, so if the implementation of a property changes that can be updated in one place. More generally, I really liked how other dbus libraries let you write properties as python properties instead of intertwined with the implementations of Get and Set, unfortunately this doesn't come for free with python-dbus but it's probably not too hard to write a custom bit of abstraction for this project.

linrunner replied:
Makes sense. Done.

tbodt commented on tlp-pd.in:
```diff
@@ -0,0 +1,394 @@
+                name="org.freedesktop.DBus.Error.InvalidArgs",
+            )
+
+        if args.debug:
```
Is there a good way to not have to write an if statement on each debug-level syslog call? I see python's logging library includes a [SysLogHandler](https://docs.python.org/3/library/logging.handlers.html#sysloghandler) and it also supports configuring minimum log levels on the logger object.

linrunner replied:
I tried with logging first. The issue was, that the syslog showed "python3" as ID instead of "tlp-pd". I took me quite some time and googling to find the solution:
```
formatter = logging.Formatter(f"{os.path.basename(__file__)}: %(message)s")
handler.setFormatter(formatter)
```

tbodt commented:
> Command line utility to control the daemon via D-Bus (similar to powerprofilesctl)

I imagined that the TLP CLI would be enough, I think

linrunner committed:
Add pyrightconfig.json

linrunner committed:
Replace syslog with logging, SysLogHandler

linrunner committed:
Make pyright happy

linrunner committed:
Set syslog identifier to "tlp-pd"

linrunner committed:
comments

linrunner committed:
message

linrunner committed:
Check root privileges

linrunner committed:
Complete introspection

linrunner committed:
Get: add properties ActionsInfo, BatteryAware

linrunner committed:
GetAll: use _get_property()

linrunner committed:
Set: add property BatteryAware

linrunner committed:
Set: use _set_property()

linrunner committed:
GetAll, Get(Profiles): make use of list and dict comprehensions

linrunner committed:
Customize polkit policy for D-Bus

TLP-PD's actions can only be performed by logged-in users with an active,
unlocked session.

Reference:
* https://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html

linrunner committed:
pyrightconfig.json: exclude worktree*/

linrunner committed:
Set(ActiveProfile): check polkit authorization

linrunner committed:
Remember bus for polkit auth

linrunner committed:
Implement a HoldProfile/ReleaseProfile list(stack)

* HoldProfile: append the new hold to the end of the list
  and activate its profile
* ReleaseProfile: hold[cookie] is removed from the list,
  then the profile is reverted to:
   - either the previous held profile (end of the list)
   - or the last profile selected by the user (list empty)
* Set(ActiveProfile): drops the whole hold list
  then sets the new profile

power-profiles-daemon works the same way.

Todo: implement the cookies in a safe, unpredictable manner.

linrunner committed:
polkit: explain policy

linrunner committed:
Sort imports

linrunner committed:
Sanitize all string input from D-Bus

linrunner committed:
comments

linrunner committed:
pyrightconfig.json: add some reportUnused

linrunner commented:
> > Command line utility to control the daemon via D-Bus (similar to powerprofilesctl)

> I imagined that the TLP CLI would be enough, I think

The D-Bus interface offers the opportunity to switch profiles without root privilege. And sooner or later, users will want it. It would also be great as a test tool, sending D-Bus messages to the daemon. 

By the way, I changed my mind and am now planning to integrate tlp-pd into the TLP repo. The Makefile is already structured  appropriately as a precaution.

It will then be added as the third subpackage tlp-pd to tlp and tlp-rdw.

I just need to discuss it with the Debian maintainer as my sparring partner.

linrunner commented:
Remaining open items:
* README, manpages
* Command line utility to control the daemon via D-Bus
  (similar to powerprofilesctl)

linrunner committed:
Instrospect: do not hardcode the whole XML, just insert the missing
property declarations

linrunner committed:
Arrange the signals according to interfaces

linrunner committed:
Introspection: fix interface for legacy object path

linrunner committed:
fix comment

linrunner committed:
Move tlp-pd to /usr/lib/

Makefile: new parameter TLP_LIB

linrunner committed:
Add manpage and README

linrunner committed:
tlp-pd: sync header text with manpage

linrunner committed:
PropertiesChanged: do not sanitize non-string input

linrunner committed:
_sanitize_input(): change replacement char to '¿'

linrunner committed:
Interface org.freedesktop.DBus.Peer: implement Ping, GetMachineId

Note: for reasons still unknown, the parent class methods are called directly.
Therefore, the declarations for this interface only provide introspection.

linrunner commented:
README, manpage done.

linrunner committed:
Get rid of tlp zombie subprocesses

linrunner committed:
Move tlp-pd back to /usr/sbin

Rationale:
* tlp-pd is an executable, not a library module
* Makes Debian's dh_python3 recognize it and silences lintian

linrunner committed:
tlp-pd: change shebang to #!/usr/bin/python3

linrunner committed:
tlp-pd: catch non-existent /run/tlp/last_pwr

linrunner committed:
tlp-pd: reinstate main()

linrunner committed:
tlp-pd: silence unused-variable (Ruff F481) warnings

linrunner committed:
tlp-pd: unify log and D-Bus exception messages, add method names

linrunner commented:
@tbodt Integration of tlp-pd into the TLP repo is done now.

linrunner committed:
README: notify of archiving
